text/0106-artifact-indices.md

```diff
@@ -0,0 +1,175 @@
+- **con**: For disjoint / unique file names, the index will grow very large.
+- **con**: Needs a `Release` object as foreign key.
+
+## Artifact Bundle
+
+Built as the successor of Release Bundles, this has a separate index for `DebugId`s and supports expiration via database
+partitioning.
+
+- **pro**: Does not require a `Release` object.
+- **pro**: Has a way to expire and remove unused bundles.
+- **con**: Does not maintain an index and has limited support more than N bundles per release.
+
+# Supporting Data
+
+- We have seen customers that upload more than 200+ Bundles per release.
+- We have also seen customers that upload up to 20.000 individual files per release.
+
+One assumption that we have is that most of the files inside of bundles are never used.
+
+- A big part of the SourceMap processing time spent in Symbolicator is related to accessing / decompressing individual
+  files from bundles.
+
+# Options Considered
+
+It is crucial that we support the use-cases explained above, as the move to artifact bundles and processing in Symbolicator
+has introduced regressions.
+
+## Use existing Artifact Indexes
+
+Early iterations of the `artifact-lookup` API made use of artifact indexes. This was later removed as the API endpoint
+was not supposed to process large files. Downloading the artifact index to the API server, and performing lookups in
+it was deemed to cause problems.
+
+It was then replaced by just returning the _newest_ `N` release bundles. This has caused extreme load on the Database.
+After investigation, we found out that sorting these release bundles by date was two orders of magnitude slower than
+just returning `N` bundles unsorted. _Unsorted_ in this case would mean primary key order, which effective means
+returning the _oldest_ `N` bundles.
+
+Adding back support for artifact index lookups would temporarily solve the regression in SourceMap processing. However
+this is based on release bundles and artifact indices which are in the process of being phased out.
+
+- **pro**: Would allow querying / accessing the needed files.
+- **con**: Requires downloading / querying the index on API servers.
+- **con**: Based on deprecated technology that is in the process of being phased out.
+
+## Using a replacement flat-file index
+
+This would involve creating a replacement for artifact indexes based on the newer artifact bundles. Instead of a `JSON`
+index, a flat-file format was proposed. However the exact format is rather an insignificant detail.
+
+- **pro**: Would allow querying / accessing the needed files.
+- **pro**: Would solve one pain-point of artifact indexes: No more `Release` object needed.
+- **con**: Requires downloading / querying the index on API servers.
+- **con**: Needs to re-create / replicate most of the artifact-index logic, and inherits all of its cons:
+  - updating / merging the index on uploads
+  - the index growing monotonically on disjoint uploads
```

Comment:
```suggestion
- **pro**: Would allow querying / accessing the needed files.
- **pro**: Would solve one pain-point of artifact indexes: No more `Release` object needed.
- **con**: Requires downloading / querying the index on API servers.
- **con**: Needs to re-create / replicate most of the artifact-index logic, and inherits all of its cons:
  - updating / merging the index on uploads
  - the index growing monotonically on disjoint uploads
- **con**: Maintenance of the index can be tedious due to automatic deletions (e.g., what to do if the bundle mentioned in the flat-file is expired?).
```